rfcs/2023-09-19-18604-metrics-data-model-refactor.md

```diff
@@ -0,0 +1,466 @@
+  the output of that codec
+
+## Scope
+
+### In scope
+
+- Adding/removing metric types to the data model.
+- Updating metrics-capable components to handle changes to the metrics data model.
+
+### Out of scope
+
+- Bolstering supporting for handling/manipulating metrics in a pipeline (e.g. enhancing metrics
+  support in the `lua` transform).
+- Explicitly adding support for other metrics systems (e.g. OpenTelemetry).
+- Deeper changes to the metrics data model such as encoding aggregation temporality (i.e.
+  `MetricKind`) into metric types directly.
+
+## Pain
+
+While the current metrics data model is, on the surface, scoped well to the metrics systems we
+support, its specificity often leads to issues with transforming metrics from one metrics to
+another.
+
+### Some metric types are specific to a single metrics system
+
+An example of this is Prometheus and Vector's native support for aggregated histograms and
+summaries. In Prometheus, aggregated histograms and summaries represent a point-in-time snapshot of
+a distribution of values, encoded into either buckets or summarized by the probability distribution.
+Both of these data types, when shown in a human-readable form, are essentially a group of gauges
+where some gauges represent summary statistics (total sum of values, total count of values) and the
+rest represent the individual buckets or quantiles of the distribution.
+
+While these data types allow us to easily ingest Prometheus data, as well as emit it, problems arise
+when transforming these metrics to other metrics systems. In particular, many metrics systems have
+no concept at all of an "aggregated histogram". Perhaps the system only supports counters, gauges,
+and histograms (but aggregated server-side, so Vector only sends the raw samples) and so the
+aggregated metrics have to be decomposed... or perhaps the aggregated histograms can be transformed
+to something like [DDSketch](https://www.vldb.org/pvldb/vol12/p2195-masson.pdf). Whatever the
+mechanism, it's an additional metric type that must be supported individually for all metrics sinks.
+
+### Needless differentiation within metric types
+
+An example of this is the inherent "statistics" for a distribution metric.
+
+In practice, this leads to Vector users being forced to choose/consider the difference between a
+"histogram" and a "summary", when internally, there is no actual difference: a distribution carries
+raw sample values, and merely calculate summary statistics over those sample values. For metric
+sinks, those summary statistics are never used, as we simply deal with the raw sample values.
+
+Beyond the user experience pain, this is extra code that needs to be handled/dealt with internally,
+even when (again) it has no inherent value.
+
+## Proposal
+
+We would make the following changes to the metrics data model:
+
+- remove `MetricValue::AggregatedSummary` and `MetricValue::Set` entirely
```

Comment:
The entire discussion up to this point is about distribution statistics. I'm not seeing the motivating pain behind removing the `Set` metric value. I see some rationale discussed in the implementation section below, but expected more about what problem it is causing. Note that I do not specifically object to the removal of the `Set` value, as it is somewhat weird compared to all the other metric values.

Reply:
The proposal to remove `Set` isn't a large, ongoing pain, but one along the lines of aggregated summaries: we _can_ handle them, but every sink needs to do so over and over again.

Essentially, if we're constantly doing the same exact thing (aggregating) in the sinks _but_ we also don't see sets being used empirically, is it worth carrying around that boilerplate and having to repeat it over and over, time and time again, for every new metrics sink?